### PR TITLE
fix(cli): parse JSON output

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
Alternative approach for #770, I think I like it better. I strongly typed every possible message mentioned in https://www.terraform.io/docs/internals/machine-readable-ui.html (thank you @jsteinich for pointing me to it).

### CODE REVIEW

**Comment by jsteinich** [comment]
We'll need to supply the `-json` argument to the Terraform cli commands. Hopefully this is just ignored in versions < 0.15.3, but if not, will need to omit.

We also probably aren't yet at a point where we can drop support for < 0.15.3, so either need to be explicit about when to parse JSON or try always and fall back to old behavior on failure.

File: packages/cdktf-cli/package.json [hunk 1]
```diff
@@ -52,7 +52,8 @@
     "stream-buffers": "^3.0.2",
     "strip-ansi": "^6.0.0",
     "uuid": "^8.3.0",
-    "yargs": "^15.1.0"
+    "yargs": "^15.1.0",
+    "zod": "^1.11.7"
```
Comment (ansgarm) [inline packages/cdktf-cli/package.json:56]: That lib looks nice! Is there a reason to not use v2 or v3?

Comment (ansgarm) [reply packages/cdktf-cli/package.json:56]: Ah found out why myself: Because v3 requires TypeScript 4.1+ and v2 is discontinued 😄

Comment (skorfmann) [reply packages/cdktf-cli/package.json:56]: Perhaps we should upgrade Typescript - for the CLI that shouldn't be a problem.

**Comment by skorfmann** [comment]
> We'll need to supply the `-json` argument to the Terraform cli commands. Hopefully this is just ignored in versions < 0.15.3, but if not, will need to omit.

Looks like the cli fails prior to `0.15.3`.

> We also probably aren't yet at a point where we can drop support for < 0.15.3, so either need to be explicit about when to parse JSON or try always and fall back to old behavior on failure.

I think we should be rather aggressive here. Probably with `0.5` we should make a cut and support Terraform `1.0` only. Since that's affecting the CLI only, users still have the option to fallback to an older Terraform CLI by using it directly.

**Comment by skorfmann** [comment]
> Alternative approach for #770, I think I like it better. I strongly typed every possible message mentioned in

That's pretty cool 👍 

The tests are failing due to the log statements. Is there anything else which is to do here?

**Comment by jsteinich** [comment]
> I think we should be rather aggressive here. Probably with `0.5` we should make a cut and support Terraform `1.0` only. Since that's affecting the CLI only, users still have the option to fallback to an older Terraform CLI by using it directly.

I'm a bit torn here. I understand not wanting to have baggage, but Terraform has been moving through versions quickly (0.15 came out 2 months ago, 0.14 was 6 months ago). I personally have stacks created with cdktf that are on 0.13.x.

Module are provider versions are always compatible across Terraform versions, so this can bleed outside of just the cli.

**Comment by DanielMSchmidt** [comment]
> The tests are failing due to the log statements. Is there anything else which is to do here?

Mainly debug the failing tests, if it's just the logs I'll remove them now 👍

**Comment by skorfmann** [comment]
> > I think we should be rather aggressive here. Probably with `0.5` we should make a cut and support Terraform `1.0` only. Since that's affecting the CLI only, users still have the option to fallback to an older Terraform CLI by using it directly.
> 
> I'm a bit torn here. I understand not wanting to have baggage, but Terraform has been moving through versions quickly (0.15 came out 2 months ago, 0.14 was 6 months ago). I personally have stacks created with cdktf that are on 0.13.x.

Is there any blocker preventing you from upgrading?

**Comment by jsteinich** [comment]
> Is there any blocker preventing you from upgrading?

Time

**Comment by skorfmann** [comment]
> > Is there any blocker preventing you from upgrading?
> 
> Time

Hehe, ok :) 

Right now it works for both output versions (text / json). At latest with Terraform 1.1, we're going to drop support for 0.15 anyway. However, I'd still push for an earlier move in that direction. And again, this only affects the `cdktf-cli` anyway, `deploy`, `diff` and `destroy` particularly

**Pull request merged**